Help on empty pages

jorinvo opened the pull request:
This should solve #128.

I did a couple of changes which I think help users when ending up on those empty pages:

- Help user when package has no built versions

  Ask user which version they are looking for and redirect them to the
  appropriate page.

  First I thought about simply explaining which URL the user has to go to
  but this seems friendlier.  
  My second idea was to do the redirect in JS on the client directly, but
  I decided that the extract work is worth it to not depend on JS.

  redirect-to-version is similar to jump-to-project. Maybe we can find
  some way to combine them?

  Also I reused the validation of request-build. Not sure if that's the
 right way.

- Hide "Other artifacts under" if group is empty

- Render "No known artifacts under the group" 

- Add top-bar with home link to empty pages

This is how the group, package and version page would look like:

http://localhost:8000/d/org.clojure
![group](https://user-images.githubusercontent.com/738978/46519708-9fa83880-c879-11e8-84a7-0c498a9e8d91.png)

http://localhost:8000/d/org.clojure/tools.logging
![package](https://user-images.githubusercontent.com/738978/46519719-a9ca3700-c879-11e8-91de-f456b7c01eda.png)

http://localhost:8000/d/org.clojure/tools.logging/1.2.3
![version](https://user-images.githubusercontent.com/738978/46519722-af278180-c879-11e8-9f74-a8500524cf7a.png)

Please let me know what you think and what I could improve!
Also, please have a look at the code. I am still learning about the Clojure ways of doing things :)

jorinvo commented on src/cljdoc/render/layout.clj:
```diff
@@ -84,27 +84,35 @@
-       [:span.dib (util/gh-coordinate scm-url)]]
-      [:a.f6.link.blue {:href (util/github-url :userguide/scm-faq)} "SCM info missing"])]])
+(defn top-bar
+  ([]
```
I added this function overload. Not sure if there is a more idiomatic way.

martinklepsch replied:
I think an alternative approach could have been to create a separate function `top-bar-generic` (since the difference is that it's not specific to a particular project). 

This would also keep the diff smaller making it potentially easier to follow for reviewers.

jorinvo replied:
Sure, I can do that! Should I share the markup or is the duplication alright?

jorinvo commented on src/cljdoc/server/routes.clj:
```diff
@@ -39,6 +39,7 @@
 
 (defn utility-routes []
   #{["/jump/release/*project" :get nop :route-name :jump-to-project]
+    ["/redirect-to-version" :post nop :route-name :redirect-to-version]
```
Really not sure about the naming of the route. And should it be placed under `/jump`?

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -32,22 +32,28 @@
+          [:div.pa4-ns.pa2
+           [:h1 group-id]
+           (if (empty? artifacts)
+             [:span.db "No known artifacts under the group " group-id]
```
sounds good!

jorinvo commented:
Totally understand your point here. Having a link to the current version would be way more helpful. I am just not sure where to get the current version from. I also know that it will only work for Maven packages once #85 is done. Do you have any points to where I can find the current version?

Also, do you prefer if I keep the commit history for these attempts here or should I amend/cherry-pick for only the final version of the PR?

martinklepsch commented:
> Do you have any points to where I can find the current version?

Sorry I should have mentioned that 🙈 😄 There is a "pseudo version" `CURRENT` that can be used like this:

```
https://cljdoc.xyz/d/org.clojure/tools.logging/CURRENT 
```

Also for context: the search on the front-page only searches Clojars because Clojars exposes a JSON-based search API. The CURRENT mechanism does not rely on this search API and instead talks to the Maven repository directly (Clojars and Maven Central can both be considered Maven repositories). See [`cljdoc.util.repositories`](https://github.com/cljdoc/cljdoc/blob/master/src/cljdoc/util/repositories.clj) if you want to better understand what's going on.

> Also, do you prefer if I keep the commit history for these attempts here or should I amend/cherry-pick for only the final version of the PR?

No strong opinions here, feel free to do what's easiest for you. I might squash/rearrange commits while merging if it makes sense (e.g. when a commit has mostly been reverted).

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -56,28 +62,43 @@
-                  {:href (routes/url-for :artifact/index :path-params (assoc cache-id :artifact-id a))}
-                  a]])]])]
+        artifacts (:artifacts cache-contents)
+        versions (->> (:versions cache-contents)
```
Extract `artifacts` and `versions` in let

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -56,28 +62,43 @@
+                      (filter #(= (:artifact-id %) (:artifact-id route-params)))
+                      (sort-by :version v/version-compare)
+                      (reverse))
+        btn-link :a.dib.bg-blue.white.ph3.pv2.br1.no-underline.f5.fw5
```
Button styles, please feel free to make it prettier!

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -56,28 +62,43 @@
+        btn-link :a.dib.bg-blue.white.ph3.pv2.br1.no-underline.f5.fw5
+        big-btn-link :a.db.link.blue.ph3.pv2.bg-lightest-blue.hover-dark-blue.br2
+        ]
+    (->> [:div
```
Wrapped in a div to include `(layout/top-bar-generic)`.

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -56,28 +62,43 @@
+          (layout/top-bar-generic)
+          [:div.pa4-ns.pa2
+           [:h1 (util/clojars-id artifact-entity)]
+           (if (empty? versions)
```
Render link on empty versions

jorinvo commented on src/cljdoc/renderers/html.clj:
```diff
@@ -56,28 +62,43 @@
+                  [big-btn-link
+                   {:href (routes/url-for :artifact/version :path-params (merge cache-id v))}
+                   (:version v)]])]])
+           (when-not (or (empty? artifacts) (= #{artifact-id} (set artifacts)))
```
Hide when artifacts empty

jorinvo commented:
I refactored my code to use `top-bar-generic`, adapted the message on the group page and replaced the form + API endpoint with the way simpler solution of linking to the current version:


![screen shot 2018-10-06 at 14 21 00](https://user-images.githubusercontent.com/738978/46571387-643e6480-c974-11e8-8896-ef19c18e6186.png)

I tried commenting the parts I changed since it is rather difficult to see in old this markup which parts actually changed. Please let me know, if there is anything else to improve!

martinklepsch commented:
Thanks for this @jorinvo, and congrats on your <strike>first</strike>, err, second! contribution to cljdoc 🎉 🍻 

I decided to squash the commits because the first few commits have been revised quite a bit.

I also used the pretty awesome [Autochrome](https://github.com/ladderlife/autochrome) to get a more readable diff for this change and it worked very nicely. You can see [the diff here](https://cdn.rawgit.com/martinklepsch/59642160207e4cb1657fda3a68a0c793/raw/d2ab166bc0282662f8d31eb79cdfa4de25d99bea/cljdoc-131-diff.html).
